### PR TITLE
intro and use fix_permissions() helper

### PR DESCRIPTION
**Pull request by mstormi**
we need to modularize this to work on different OS (and with and without ZRAM)

Fixes: #1369 

Signed-off-by: Markus Storm <markus.storm@gmx.net>

### CODE REVIEW

File: functions/system.bash [hunk 1]
```diff
@@ -302,19 +302,24 @@ permissions_corrections() {
...
   fi
 
-  if ! cond_redirect chown --recursive "${username:-openhabian}:${username:-openhabian}" "/home/${username:-openhabian}"; then echo "FAILED (${username:-openhabian} own $HOME)"; return 1; fi
+  if ! cond_redirect fix_permissions  "/home/${username:-openhabian}" "${username:-openhabian}:${username:-openhabian}"; then echo "FAILED (${username:-openhabian} chown $HOME)"; return 1; fi
   
-  if [ -f /etc/mosquitto/passwd ]; then
-    if ! cond_redirect chown --silent "mosquitto:${username:-openhabian}" /etc/mosquitto/passwd /opt/zram/log.bind/mosquitto; then echo "FAILED (mosquitto passwd and log permissions)"; return 1; fi
-    if ! cond_redirect chmod 660 /etc/mosquitto/passwd; then echo "FAILED (mosquitto passwd permissions)"; return 1; fi
-  fi
+  if ! cond_redirect fix_permissions /etc/mosquitto/passwd "mosquitto:${username:-openhabian}" 660 770; then echo "FAILED (mosquitto passwd permissions)"; return 1; fi
```
Comment (ecdye) [inline functions/system.bash:307]: Don't forget to check if Mosquitto is installed before running this.

File: functions/system.bash [hunk 2]
```diff
@@ -302,19 +302,24 @@ permissions_corrections() {
...
+  if ! cond_redirect fix_permissions /opt/zram/log.bind/mosquitto "mosquitto:${username:-openhabian}" 664 775; then echo "FAILED (mosquitto log permissions on ZRAM)"; return 1; fi
   if ! cond_redirect setfacl --recursive --remove-all "${openhabFolders[@]}"; then echo "FAILED (reset file access lists)"; return 1; fi
   # not sure if still needed. Let's see if removing this causes any user issues
   #if ! cond_redirect setfacl -R -m g::rwX "${openhabFolders[@]}"; then echo "FAILED (set file access)"; return 1; fi
   #if ! cond_redirect setfacl -R -m d:g::rwX "${openhabFolders[@]}"; then echo "FAILED"; return 1; fi
 
-  if cond_redirect chgrp --silent root /var/log/samba /var/log/unattended-upgrades /opt/zram/log.bind/samba; then echo "OK"; else echo "FAILED (samba and 3rd party logdir)"; return 1; fi
-  if cond_redirect chown --silent --recursive "${username:-openhabian}:${username:-openhabian}" /opt/zram/persistence.bind/ /opt/zram/log.bind/openhab; then echo "OK"; else echo "FAILED (persistence)"; return 1; fi
+  if cond_redirect fix_permissions /var/log/unattended-upgrades root:root 664 755; then echo "OK"; else echo "FAILED (unattended upgrades logdir)"; return 1; fi
```
Comment (ecdye) [inline functions/system.bash:315]: I have seen issues with this Directory not existing before, please check that it exists before running.

File: functions/system.bash [hunk 3]
```diff
@@ -302,19 +302,24 @@ permissions_corrections() {
...
+  if cond_redirect fix_permissions /var/log/samba root:root 660 770; then echo "OK"; else echo "FAILED (samba logdir)"; return 1; fi
+  if cond_redirect fix_permissions /opt/zram/log.bind/samba root:root 660 770; then echo "OK"; else echo "FAILED (samba logdir on ZRAM)"; return 1; fi
+
+  if cond_redirect fix_permissions /opt/zram/persistence.bind "openhab:${username:-openhabian}" 664 775; then echo "OK"; else echo "FAILED (persistence)"; return 1; fi
+
+  if cond_redirect fix_permissions /var/log/openhab "openhab:${username:-openhabian}" 664 775; then echo "OK"; else echo "FAILED (openhab log)"; return 1; fi
+  if cond_redirect fix_permissions /opt/zram/log.bind/openhab "openhab:${username:-openhabian}" 664 775; then echo "OK"; else echo "FAILED (openhab log on ZRAM)"; return 1; fi
```
Comment (ecdye) [inline functions/system.bash:322]: For Code readability, can you only add the blank lines between each group of folders/files.

File: functions/helpers.bash [hunk 4]
```diff
@@ -526,3 +526,24 @@ zram_dependency() {
   if ! cond_redirect systemctl -q daemon-reload &> /dev/null; then echo "FAILED (daemon-reload)"; return 1; fi
 }
 
+## fix permissions on a file OR recursively on a directory that MIGHT be on zram
```
Comment (ecdye) [inline functions/helpers.bash:529]: Capitalize first letter.

**Review by ecdye — CHANGES_REQUESTED** [review]
I'm still not sure why the lower-dir permissions need to be changed but it should not hurt.

**Comment by mstormi** [comment]
> I'm still not sure why the lower-dir permissions need to be changed but it should not hurt.

my thinking too. If for some reason zram isn't up by the time the permissions on lowerFS are changed this might remain and fall back that status when zram is shutdown.

File: functions/system.bash (hunk 2 above)
Comment (mstormi) [reply functions/system.bash:315]: the 1st command in fix_permissions() is always [[ -e ]] to check the file or dir and return 0 if it does not

File: functions/system.bash (hunk 1 above)
Comment (mstormi) [reply functions/system.bash:307]: done

File: functions/system.bash (hunk 3 above)
Comment (mstormi) [reply functions/system.bash:322]: I intentionally did ... it's a question of grouping isn't it

File: functions/helpers.bash (hunk 4 above)
Comment (mstormi) [reply functions/helpers.bash]: ok

**Comment by mstormi** [comment]
ok as I believe I addressed all your comments and have positive test feedback in #1369 I think I'm going to merge it

So let's keep the issues on permissions coming you know where to look after ;-)

**Pull request merged**